### PR TITLE
Group workspace masking error in MaskDetectors

### LINKED ISSUE #14793: Masking using a loaded cal file with grouping does not work
This works

``` python
data = Load(Filename=r'\\isis\inst$\ndxwish\instrument\data\cycle_15_3\WISH00032918.nxs')
groupws = LoadCalFile(InputWorkspace = data,
CalFilename='C:/Users/rrc79113/Desktop/32918_165nm_down_rot180_upper.cal', 
WorkspaceName='cal', MakeOffsetsWorkspace = False, MakeMaskWorkspace = False)

detlist = []
for i in range(groupws.getNumberHistograms()):
    #if group 0 - discard
    if groupws.dataY(i)[0] == 0:
        det = groupws.getDetector(i)
        detlist.append(det.getID())
    else:
        det = groupws.getDetector(i)
        print "not masking {0}".format(det.getID())
print "found {0} detectors to mask".format(len(detlist))
MaskDetectors(Workspace=data,DetectorList=detlist)
```

but this doesn't and it should 

``` python
data = Load(Filename=r'\\isis\inst$\ndxwish\instrument\data\cycle_15_3\WISH00032918.nxs')
maskws = LoadCalFile(InputWorkspace = data,
CalFilename='C:/Users/rrc79113/Desktop/32918_165nm_down_rot180_upper.cal', 
WorkspaceName='cal', MakeOffsetsWorkspace = False, MakeGroupingWorkspace = False)
MaskDetectors(Workspace=data,MaskedWorkspace=maskws)
```

 it fails with this error

Error in execution of algorithm MaskDetectors:
IndexError: MatrixWorkspace::maskWorkspaceIndex,index 194565 :: 0 <==> 194565

### PR DESCRIPTION
**Pull request by samueljackson92**
See the two scripts in the bug report for examples of how to reproduce the issue.

**To test:**
Two cases to check that didn't work before:
 - Check that a workspace (NOT a masked workspace) where
    - masked workspace has a number of spectra == the number of detectors in the input workspace
    - masked workspace has a number of spectra > the number of spectra in the input workspace
 - Check the above works when specifying a `StartWorkspaceIndex` and `EndWorkspaceIndex`
 - Code review.

Fixes #14793 .

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant documentation been added/updated?
- [x] Is user-facing documentation written in a user-friendly manner?
- [x] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/DataHandling/src/MaskDetectors.cpp [hunk 1]
```diff
@@ -172,17 +172,35 @@ void MaskDetectors::exec() {
...
       // Check the provided workspace has the same number of spectra as the
       // input
-      if (prevMasking->getNumberHistograms() > WS->getNumberHistograms()) {
-        g_log.error() << "Input workspace has " << WS->getNumberHistograms()
-                      << " histograms   vs. "
-                      << "Input masking workspace has "
-                      << prevMasking->getNumberHistograms()
-                      << " histograms. \n";
-        throw std::runtime_error("Size mismatch between two input workspaces.");
+      if (nHist > WS->getNumberHistograms()) {
```
Comment (NickDraper) [inline Framework/DataHandling/src/MaskDetectors.cpp]: This if does not fit the comment above it

Comment (NickDraper) [inline Framework/DataHandling/src/MaskDetectors.cpp]: We also need to consider what should happen if it is <, probably handle it this way

File: Framework/DataHandling/src/MaskDetectors.cpp [hunk 2]
```diff
@@ -172,17 +172,35 @@ void MaskDetectors::exec() {
...
+        // Check the provided workspace has the same number of detectors as the
+        // input. If so add to the detector list
+        auto instrument = WS->getInstrument();
+        auto maskInstrument = prevMasking->getInstrument();
+        if (instrument && maskInstrument &&
+            maskInstrument->getNumberDetectors() ==
+                instrument->getNumberDetectors() &&
+            nHist == instrument->getNumberDetectors()) {
```
Comment (NickDraper) [inline Framework/DataHandling/src/MaskDetectors.cpp]: I don't think we should have the number of detectors == nhist restriction.

File: Framework/DataHandling/src/MaskDetectors.cpp [hunk 3]
```diff
@@ -172,17 +172,35 @@ void MaskDetectors::exec() {
...
-                      << " histograms. \n";
-        throw std::runtime_error("Size mismatch between two input workspaces.");
+      if (nHist > WS->getNumberHistograms()) {
+        // Check the provided workspace has the same number of detectors as the
+        // input. If so add to the detector list
+        auto instrument = WS->getInstrument();
+        auto maskInstrument = prevMasking->getInstrument();
+        if (instrument && maskInstrument &&
+            maskInstrument->getNumberDetectors() ==
+                instrument->getNumberDetectors() &&
```
Comment (NickDraper) [inline Framework/DataHandling/src/MaskDetectors.cpp]: if the number of detectors in the instruments do not match we should just warn but continue, not refuse.

File: Framework/DataHandling/src/MaskDetectors.cpp [hunk 4]
```diff
@@ -172,17 +172,35 @@ void MaskDetectors::exec() {
...
+            nHist == instrument->getNumberDetectors()) {
+          appendToDetectorListFromWS(detectorList, WS, prevMasking,
+                                     ranges_info);
+        } else {
+          // We have more spectra in the mask than in the workspace and we also
```
Comment (NickDraper) [inline Framework/DataHandling/src/MaskDetectors.cpp]: So we should only refuse if one of the WS's does not have an instrument.

File: Framework/DataHandling/src/MaskDetectors.cpp [hunk 5]
```diff
@@ -172,17 +172,35 @@ void MaskDetectors::exec() {
...
+          // give up loudly.
+          g_log.error() << "Input workspace has " << WS->getNumberHistograms()
+                        << " histograms   vs. "
+                        << "Input masking workspace has "
+                        << prevMasking->getNumberHistograms()
+                        << " histograms. \n";
+          throw std::runtime_error(
+              "Size mismatch between two input workspaces.");
+        }
+      } else {
```
Comment (NickDraper) [inline Framework/DataHandling/src/MaskDetectors.cpp]: This else should be more explicit of when it is being called (when workspace indices match)

File: Framework/DataHandling/src/MaskDetectors.cpp [hunk 6]
```diff
@@ -454,8 +472,38 @@ void MaskDetectors::appendToIndexListFromWS(
...
+    const std::tuple<size_t, size_t, bool> &range_info) {
+  size_t startIndex = std::get<0>(range_info);
+  size_t endIndex = std::get<1>(range_info);
+
+  const auto &detMap = inputWs->getDetectorIDToWorkspaceIndexMap();
+  detectorList.reserve(maskWs->getNumberHistograms());
+
+  for (size_t i = 0; i < maskWs->getNumberHistograms(); ++i) {
+    if (maskWs->y(i)[0] == 0) {
+      auto id = maskWs->getDetector(i)->getID();
```
Comment (NickDraper) [inline Framework/DataHandling/src/MaskDetectors.cpp]: There is a methods that will return a vector of all of the IDs (i think getIDs) that will allow this to work for none 1-1 mapping.

**Review by NickDraper — CHANGES_REQUESTED** [review]
#we also need some changes to MaskDetectors.rst to make it explicit what will happen when passing in a MaskedWorkspace, as there are several options.

**Comment by samueljackson92** [comment]
@NickDraper Please review again when you have time.

File: Framework/DataHandling/src/MaskDetectors.cpp [hunk 7]
```diff
@@ -260,6 +217,82 @@ void MaskDetectors::exec() {
...
+  const auto nHist = maskWs->getNumberHistograms();
+  auto instrument = WS->getInstrument();
+  auto maskInstrument = maskWs->getInstrument();
+
+  // Check the provided workspace has the same number of spectra as the
+  // input, if so assume index list
+  if (nHist == WS->getNumberHistograms()) {
+    appendToIndexListFromWS(indexList, maskWs, rangeInfo);
+    // Check they both have instrument and then use detector based masking
+  } else if (instrument && maskInstrument) {
```
Comment (NickDraper) [inline Framework/DataHandling/src/MaskDetectors.cpp]: This should check the instrument->name, not pointer address, but it also should only log as warning if they do not match, still go ahead and apply the masking.

File: Framework/DataHandling/src/MaskDetectors.cpp [hunk 8]
```diff
@@ -260,6 +217,82 @@ void MaskDetectors::exec() {
...
+    const auto inputDetCount = instrument->getNumberDetectors();
+    const auto maskDetCount = maskInstrument->getNumberDetectors();
+    if (inputDetCount != maskDetCount)
+      g_log.warning() << "Number of detectors does not match between "
+                         "mask workspace and input workspace";
+
+    appendToDetectorListFromWS(detectorList, WS, maskWs, rangeInfo);
```
Comment (NickDraper) [inline Framework/DataHandling/src/MaskDetectors.cpp]: We should log at info that the masking is being applied by detector ids

File: Framework/DataHandling/src/MaskDetectors.cpp [hunk 9]
```diff
@@ -260,6 +217,82 @@ void MaskDetectors::exec() {
...
+    const RangeInfo &rangeInfo) {
+
+  const auto nHist = maskWs->getNumberHistograms();
+  auto instrument = WS->getInstrument();
+  auto maskInstrument = maskWs->getInstrument();
+
+  // Check the provided workspace has the same number of spectra as the
+  // input, if so assume index list
+  if (nHist == WS->getNumberHistograms()) {
+    appendToIndexListFromWS(indexList, maskWs, rangeInfo);
```
Comment (NickDraper) [inline Framework/DataHandling/src/MaskDetectors.cpp]: We should log at info that the masking is being applied by workspace index

File: Framework/DataHandling/src/MaskDetectors.cpp [hunk 10]
```diff
@@ -260,6 +217,82 @@ void MaskDetectors::exec() {
...
+  }
+
+  g_log.debug() << "Extracting mask from MaskWorkspace (" << maskWs->getName()
+                << ")\n";
+  bool forceDetIDs = getProperty("ForceInstrumentMasking");
+  if (maskWs->getNumberHistograms() != WS->getNumberHistograms() ||
+      forceDetIDs) {
+    extractMaskedWSDetIDs(detectorList, maskWs);
+  } else {
+    appendToIndexListFromMaskWS(indexList, maskWs, rangeInfo);
```
Comment (NickDraper) [inline Framework/DataHandling/src/MaskDetectors.cpp]: We should log at info that the masking is being applied by workspace index

File: Framework/DataHandling/src/MaskDetectors.cpp [hunk 11]
```diff
@@ -260,6 +217,82 @@ void MaskDetectors::exec() {
...
+    throw std::runtime_error("Instrument's detector numbers mismatch "
+                             "between input Workspace and MaskWorkspace");
+  }
+
+  g_log.debug() << "Extracting mask from MaskWorkspace (" << maskWs->getName()
+                << ")\n";
+  bool forceDetIDs = getProperty("ForceInstrumentMasking");
+  if (maskWs->getNumberHistograms() != WS->getNumberHistograms() ||
+      forceDetIDs) {
+    extractMaskedWSDetIDs(detectorList, maskWs);
```
Comment (NickDraper) [inline Framework/DataHandling/src/MaskDetectors.cpp]: We should log at info that the masking is being applied by detector ids

**Review by NickDraper — CHANGES_REQUESTED** [review]

**Review by NickDraper — APPROVED** [review]

**Comment by NickDraper** [comment]
Tested using the following script 
``` python
wsUnmasked = CreateSampleWorkspace()
wsMasked = CloneWorkspace(wsUnmasked)

def GeneratePattern(ws,noToGroup):
    groupingPattern = []
    for i in range(0,ws.getNumberHistograms(),noToGroup):
         groupingPattern.append(str(i) + '+' + str(i+1))
    groupingPatternString = ", ".join(groupingPattern)
    return groupingPatternString
    
wsGrouped2 = GroupDetectors(wsUnmasked,GroupingPattern = GeneratePattern(wsUnmasked,2))

# Define an infinite cylinder with its axis parallel to the Z-axis
# and the radius of 0.04
shapeXML = \
"""
<infinite-cylinder id="A" >
    <centre x="0" y="0" z="0" />
    <axis x="0" y="0" z="1" />
    <radius val="0.04" />
</infinite-cylinder>
"""
# Mask all detectors inside this cylinder
wsMasked = CloneWorkspace(wsUnmasked)
MaskDetectorsInShape( wsMasked, shapeXML )


wsGrouped2Masked = CloneWorkspace(wsGrouped2)
MaskDetectorsInShape( wsGrouped2Masked, shapeXML )


wsTestGrouped2MaskCopy = CloneWorkspace(wsUnmasked)
MaskDetectors(wsTestGrouped2MaskCopy,MaskedWorkspace=wsGrouped2Masked)

wsTestGroupedMaskCopy = CloneWorkspace(wsGrouped2)
MaskDetectors(wsTestGroupedMaskCopy,MaskedWorkspace=wsMasked)
```

Then the instrument views of the workspaces were compared.

Seems to work well :shipit:

**Comment by martyngigg** [comment]
@samueljackson92 There was a trivial conflict in the release notes. I've resolved it in the browser and I'll merge this assuming the builds pass.

**Comment by martyngigg** [comment]
Failing Windows test is unrelated slice viewer failure.

**Pull request merged**